rfd/0166-tpm-joining.md

```diff
@@ -0,0 +1,390 @@
+---
+authors: @strideynet (noah@goteleport.com)
+state: draft
+---
+
+# RFD 0166 - TPM Joining
+
+## Required Approvers
+ 
+* Engineering: (@zmb33 || @codingllama)
+* Security: (@reedloden || @jentfoo)
+* Product: (@xinding33 || @klizhentas)
+
+# What
+
+Introduce a new join method that allows a Bot or Agent running on a host with a
+TPM to securely join a Teleport cluster without using a shared secret.
+
+# Why
+
+Bootstrapping trust with a newly provisioned host in an on-premises environment
+is challenging. In many environments, this is done by transferring an initial
+shared secret to the host, which can then be used for authentication. However,
+this is difficult to complete securely - especially at scale. This is due to the
+risk of impersonation/interception, an attacker who has sufficiently compromised
+the network can impersonate a newly provisioned host and receive access to this
+secret. In addition, a shared secret on the host is liable to exfiltration.
+
+A TPM provides a secure, unique and persistent initial identity ideal for
+bootstrapping trust with a host. Even across reboots, or reconfigurations, the
+TPM identity remains the same. The guarantees provided by a compliant TPM mean
+that this identity cannot be exfiltrated. This makes it a strong candidate for
+bootstrapping trust.
+
+# Implementation
+
+Reference materials:
+
+- [TCG Credential Profile EK 2.0](https://trustedcomputinggroup.org/wp-content/uploads/TCG-EK-Credential-Profile-V-2.5-R2_published.pdf): For specification of the EK and EKCert.
+- [TCG Trusted Platform Module Library](https://trustedcomputinggroup.org/resource/tpm-library-specification/): For specification of TPM commands and structures.
+- [TPM 2.0 Keys for Device identity and Attestation](https://trustedcomputinggroup.org/wp-content/uploads/TPM-2p0-Keys-for-Device-Identity-and-Attestation_v1_r12_pub10082021.pdf): For the recommended TPM 2.0 flows for device identity and attestation.
+- [RFD 0008e - Device Trust TPM Support](https://github.com/gravitational/teleport.e/blob/master/rfd/0008e-device-trust-tpm.md): For the existing TPM identity implementation in Teleport.
+
+Existing code and logic from Device Trust TPM support will be reused where
+appropriate for this feature.
+
+## Overview of TPM functionality and terminology
+
+- Trusted Platform Module (TPM): a hardware module that provides a secure root
+  of trust for a host. It can securely generate and store keys and perform
+  cryptographic operations for the host without exposing the keys.
+- Endorsement Key (EK): A key pair that is "burned in" to the TPM at the time of
```

Comment:
How does this work for vTPMs? Are the keys "fixed"? Are they the same as the actual hardware TPM of the device?

Reply:
For a vTPM, the EK shouldn't match that of any TPM on the host system. vTPM implementations do vary though - there's some that are essentially entirely in-software and running on the hypervisor which can function without a hardware TPM (albeit the secureness of these is limited).

Reply:
Gotcha. Do you think it's worth addressing and/or mentioning vTPMs in the RFD? The initial use of this might be on-prem, but it's not a stretch that someone is going to try it out on vTPMs somewhere.

Reply:
I'm happy with a mention, even if it's just to say that we don't treat it specially in any way.

Footgun is always an option ;)

Reply:
ffa89df408